crypto/elligator2.c

```diff
@@ -0,0 +1,195 @@
+/** Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions: The above copyright
+ * notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+
+#include "ed25519-donna/ed25519-donna.h"
+#include "memzero.h"
+
+#include "elligator2.h"
+
+// Returns 1 if a equals b, returns 0 otherwise.
+static int curve25519_isequal(bignum25519 a, const bignum25519 b) {
+  bignum25519 difference;
+  curve25519_sub(difference, a, b);
+  int result = 1 - curve25519_isnonzero(difference);
+  memzero(difference, sizeof(difference));
+  return result;
+}
+
+// Sets out to a if c equals 0, sets out to b if c equals 1.
+static void curve25519_cmov(bignum25519 out, const bignum25519 a,
+                            const bignum25519 b, uint32_t c) {
+  assert((int)(c == 1) | (int)(c == 0));
+
+  bignum25519 a_copy = {0}, b_copy = {0};
+  curve25519_copy(a_copy, a);
+  curve25519_copy(b_copy, b);
+  curve25519_swap_conditional(a_copy, b_copy, c);
+  memzero(b_copy, sizeof(b_copy));
+  curve25519_copy(out, a_copy);
+  memzero(a_copy, sizeof(a_copy));
+}
+
+bool map_to_curve_elligator2_curve25519(const uint8_t input[32],
+                                        curve25519_key output) {
+  // https://www.rfc-editor.org/rfc/rfc9380.html#map-to-curve25519
+  // The procedure from the above link is used, with the exception that the
+  // y-coordinate of the output point is not computed, because it is not needed.
+  bignum25519 input_bignum = {0};
+  curve25519_expand(input_bignum, (unsigned char*)input);
+
+  // c3 = sqrt(-1)
+  bignum25519 c3 = {0};
+  curve25519_set_sqrtneg1(c3);
+
+  // c2 = 2^c1 = 2^((q + 3) / 8) = c3 + 1
+  bignum25519 one = {0};
+  bignum25519 c2 = {0};
```

Comment:
c2 is not needed, please, remove it

Reply:
Done in https://github.com/trezor/trezor-firmware/pull/3435/commits/2052ce286aa69dcb7edb279d10b6456bccc43205.